polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -0,0 +1,385 @@
+//===- ArgumentPromotion.cpp - Promote by-reference arguments -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass currently attempts to peel struct members from the argument pack
+// passed to a SYCL kernel. For example given:
+//
+//  gpu.func @kernel_parallel_for(%arg0: i32) {
+//    func.call @parallel_for(%0, ...) :
+//        (memref<?x!llvm.struct<(i32, !sycl.accessor), ...)
+//    gpu.return
+//  }
+//
+// The pass modifies the call (and the callee):
+//
+//  gpu.func @kernel_parallel_for(%arg0: i32) {
+//    %int_arg = <memref to first struct member>
+//    %acc_arg = <memref to second struct member>
+//    func.call @parallel_for(%int_arg, %acc_arg, ...) :
+//        (memref<?xi32>, memref<?x!llvm.struct<(!sycl.accessor...)
+//    gpu.return
+//  }
+
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Polygeist/IR/Ops.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arg-promotion"
+#define REPORT_DEBUG_TYPE DEBUG_TYPE "-report"
+
+namespace mlir {
+namespace polygeist {
+#define GEN_PASS_DEF_ARGUMENTPROMOTION
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h.inc"
+} // namespace polygeist
+} // namespace mlir
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Represent a candidate for the argument promotion transformation.
+class Candidate {
```

Comment:
Do you think it would be better for the candidate to be a `FunctionOpInterface`, i.e., the callee? That way it will be easier to extent to callee with more than one callsite.
We can check if all its callsite satisfy the candidate requirements.